### PR TITLE
Fix OUSD storage

### PR DESCRIPTION
**Pull request by tomlinton**
Use arrays to deprecate old storage. Includes upgrade script. Had to flick some variables from private to public to allow the reset to work, but we can clean that up after.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers
  - [x] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [x] Echidna tests pass (not automated, run manually on local)

### CODE REVIEW

**Comment by franckc** [comment]
Made a few changes to the deploy script to include sending the proposal to the governor on Mainnet.

Ran the deploy on fork and here is the output of the debug task: https://gist.github.com/franckc/964b2d002d065a7c34ad57f8598164f8

**Review by franckc — DISMISSED** [review]
LGTM

**Comment by DanielVF** [comment]
I'm currently running echidna on this.

File: contracts/contracts/token/OUSD.sol [hunk 1]
```diff
@@ -34,33 +34,21 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
         uint256 rebasingCreditsPerToken
     );
 
-    // MAX_SUPPLY is chosen to guarantee applied changes to _totalSupply in
-    // changeSupply(_newTotalSupply) deviate from the value provided in
-    // _newTotalSupply by < 1
-    uint256 private constant MAX_SUPPLY = ~uint128(0); // (2^128) - 1
+    enum RebaseOptions { NotSet, OptOut, OptIn }
 
+    uint256 private constant MAX_SUPPLY = ~uint128(0); // (2^128) - 1
```
Comment (DanielVF) [inline contracts/contracts/token/OUSD.sol:39]: I have verified with remix that the value set in MAX_SUPPLY really matches the comment. (Sadly, google's built in calculator lies about this - had to drop down to `bc` on the terminal)

image

File: contracts/contracts/token/OUSD.sol [hunk 2]
```diff
@@ -34,33 +34,21 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
         uint256 rebasingCreditsPerToken
     );
 
-    // MAX_SUPPLY is chosen to guarantee applied changes to _totalSupply in
-    // changeSupply(_newTotalSupply) deviate from the value provided in
-    // _newTotalSupply by < 1
-    uint256 private constant MAX_SUPPLY = ~uint128(0); // (2^128) - 1
+    enum RebaseOptions { NotSet, OptOut, OptIn }
```
Comment (DanielVF) [inline contracts/contracts/token/OUSD.sol:37]: Thanks for re-sane-ing the order of these variables.

File: contracts/deploy/009_ousd_fix.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    log("Sending proposal to governor...");
+    await sendProposal(propResetArgs, propResetDescription);
+    log("Proposal sent.");
+  } else if (isFork) {
+    // On Fork we can send the proposal then impersonate the guardian to execute it.
+    log("Sending and executing proposal...");
+    await executeProposal(propResetArgs, propResetDescription);
+    log("Proposal executed.");
+  } else {
+    await withConfirmation(
```
Comment (DanielVF) [inline contracts/deploy/009_ousd_fix.js:78]: One day it would be nice to write some code so we can just use the same propResetArgs

**Review by DanielVF — APPROVED** [review]
Looks good to me. I've not tested it with fork, but have run tests locally.

I've also run the echidna tests successfully.

**Comment by franckc** [comment]
Added a new hardhat task "executeProposalOnFork" that will allow us to easily execute a proposal on the Fork, after a Mainnet deploy but before execution by the multi-sig. Added a step in the deploy plan for running that new task and doing some verification on the contracts.

**Review by franckc — DISMISSED** [review]

**Review by franckc — APPROVED** [review]

**Review by DanielVF — APPROVED** [review]

**Pull request merged**